### PR TITLE
[FLINK-16178][FLINK-16192][checkpointing] Clean up checkpoint metadata code and remove remaining bits of "legacy state"

### PR DESCRIPTION
**Pull request by StephanEwen**
## NOTE: This removes savepoint compatibility with Flink 1.2 !

For background, please consult [FLINK-16192](https://issues.apache.org/jira/browse/FLINK-16192)

## What is the purpose of the change

**This is part of the preparation for implementing Fault Tolerance for Operator Coordinators for the new Source Interface (FLIP-27).**

This PR contains various cleanups / simplifications / refactorings around the checkpoint metadata serialization / deserialization code, to make it easier to implement the handling of Operator Coordinator State.

...and to finally drop some really old and unused code. I felt a bit like an archaeologist digging through this ;-)

## Brief change log

The main changes are:

  - 5436317bc99830bdc21b5799fcabf8f833120932 : *[FLINK-16178][refactor] Make the versioned Checkpoint Metadata Serializers only responsible for deserialization.*

    The metadata is always serialized with the latest serializer. Deserialization uses a version-specific serializer. That is the same functionality as before, but cleanly represented in the code. The code previously pretended to be version-dynamic on serialization side as well, but still could only handle the latest version on serialization. Earlier versions failed with a `ClassCastException` or `UnsupportedMethodException`.

    Weird constructs were necessary, like a `SavepointV1Serializer implements SavepointSerializer<SavepointV2>` (mismatch between handled version and generic type).

  - 247bd519b8609610c80704bb4a7c0e5a8375494a : *[FLINK-16192][checkpointing] Remove remaining bits of "legacy state" and remove Savepoint 1.2 compatibility*

    "Legacy State" Refers to the state originally created by the old `Checkpointed` interface, before state was re-scalable. This was later replaced by `CheckpointedFunction` (and `ListCheckpointed` as the shortcut).

    This state was no longer supported since Flink 1.4. However, all 1.2 savepoints in migration tests (and some 1.3 savepoints for tests) used that "legacy state", that is why the code was retained and support for "legacy state" was hackily activated (via some static flags) for migration tests.

    Removing that code would actually not reduce any savepoint compatibility, because compatibility for 1.2 and 1.3 was anyways only given for jobs that did not use that "legacy state". However, removing this code makes all 1.2 savepoints from tests unusable.

    A [mailing list discussion](https://lists.apache.org/x/thread.html/rd36579fe754f7557da4ae1dda47820c84b21009c35e6e317671e83b0@%3Cdev.flink.apache.org%3E) suggested to drop that support, tather than elaborate re-building of these the test data for Flink 1.2 with different state type. This PR also removes a few isolated 1.3 snapshot migration tests, where the snapshots contained legacy state. There should be still enough test coverage for the checkpoint metadata serializers to not render 1.3 support under tested.

## Verifying this change

This change is already covered by existing tests, such as the savepoint tests and savepoint migration tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: **yes**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointSerializer.java [hunk 1]
```diff
@@ -52,6 +37,5 @@
 	 * @return The deserialized savepoint
 	 * @throws IOException Serialization failures are forwarded
 	 */
-	T deserialize(DataInputStream dis, ClassLoader userCodeClassLoader) throws IOException;
-
+	SavepointV2 deserialize(DataInputStream dis, ClassLoader userCodeClassLoader) throws IOException;
```
Comment (buptljy) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointSerializer.java]: Should we bind #SavepointSerializer to #SavepointV2 ? 
In this way, we have to create a new interface if we want to add #SavepointV3 in the future. Is that expected?

Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointSerializer.java]: In this design, this is expected. If we introduce a new format version of the metadata, all deserializers need to deserialize to that new version.

Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointSerializer.java]: This has always been the case in the code. Previously, the interface "pretended" to be cross version generic, but in the end all deserializers had to be updated every time a new version came. This now makes this explicit.

Comment (Jiayi-Liao) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointSerializer.java]: You're right. I was misguided by the previous interface.

Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointSerializer.java]: This change here contains something similar to your suggestion: https://issues.apache.org/jira/browse/FLINK-16259

Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointSerializer.java]: The related PR is #11202

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointSerializers.java [hunk 2]
```diff
@@ -35,7 +35,7 @@
 	private static final Map<Integer, SavepointSerializer<?>> SERIALIZERS = new HashMap<>(2);
 
 	static {
-		SERIALIZERS.put(SavepointV1.VERSION, SavepointV1Serializer.INSTANCE);
+		SERIALIZERS.put(SavepointV1Serializer.VERSION, SavepointV1Serializer.INSTANCE);
 		SERIALIZERS.put(SavepointV2.VERSION, SavepointV2Serializer.INSTANCE);
```
Comment (klion26) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointSerializers.java]: Seems we just keep the recent savepoint, I'm thinking can we replacing all `SavepointV2` with `Savepoint` directly, after the replacement, we just have one `Savepoint` class, no `SavepointVX` class, and update the `VERSION` property in `Savepoint` when upgrading.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/Checkpoints.java [hunk 3]
```diff
@@ -103,7 +102,7 @@ public static Savepoint loadCheckpointMetadata(DataInputStream in, ClassLoader c
```
Comment (klion26) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/Checkpoints.java]: maybe we can replace `Savepoint` with `SavepointV2` on line 97.

**Comment by StephanEwen** [comment]
The next step of cleanups is in #11202

**Comment by StephanEwen** [comment]
@carp84 I see you assigned yourself as a reviewer. Do you plan to give this a pass or is this okay from your side?

**Comment by StephanEwen** [comment]
@klion26 Thank you for taking a look! Your suggestions are implemented in this follow-up PR: #11202

There is only one "Savepoint" class any more (renamed to "CheckpointMetadata") and it is not versioned. Only the serializers are versions.

I would merge the two PRs together, because the second PR finishes the work of the first. I was only trying to spit it into different PRs so they would not grow too big.

**Comment by StephanEwen** [comment]
Merged in b8277f92cfd1fa80325194bd705d1a13293b30e3

**Pull request closed without merging**